lib/wallet/src/Cardano/Wallet/Primitive/AddressDerivation.hs

```diff
@@ -596,6 +606,17 @@ instance KnownNat pm => NetworkDiscriminantVal ('Staging pm) where
     networkDiscriminantVal =
         "staging (" <> T.pack (show $ natVal $ Proxy @pm) <> ")"
 
+networkVal
+    :: forall (n :: NetworkDiscriminant). Typeable n
+    => Word8
+networkVal = fromMaybe (error $ "network: unmatched type" <> show (typeRep @n))
+       tryMainnet
+  where
+    tryMainnet =
+        case testEquality (typeRep @n) (typeRep @'Mainnet) of
+            Just Refl  -> Just 0b00000001
+            Nothing -> Just 0b00000000
```

Comment:
I believe this will cause `Staging` to be incorrectly treated as `Testnet`

Reply:
A strictly better way of implementing `networkVal` should to to either add it as part of the `NetworkDiscriminantVal` class or to create a new similar class for it. Then we'd be forced to implement it for all cases [here](https://github.com/input-output-hk/cardano-wallet/blob/76f09ef7fda58418710b8394603d75d677b34425/lib/wallet/src/Cardano/Wallet/Primitive/AddressDerivation.hs#L587-L597). And we'd avoid the need for `fromMaybe (error $ "network: unmatched type" <> show`.

Reply:
no, when it comes to network tag there is one that differentiate mainnet (0) and the rest (1). Meaning, you can use mainnet addresses only in mainnet, and nonmainnet addresses elsewhere.

Reply:
I thought `Staging` was supposed to use the same addresses as `Mainnet`. And it does seems to be reflected in our `PaymentAddress` logic:

This instance for `Staging` reuses the implementation from `Mainnet`:
```haskell
instance PaymentAddress 'Mainnet k ktype => PaymentAddress ('Staging pm) k ktype where
    paymentAddress = paymentAddress @'Mainnet @k @ktype
    liftPaymentAddress = liftPaymentAddress @'Mainnet @k @ktype
```
https://github.com/input-output-hk/cardano-wallet/blob/master/lib/wallet/src/Cardano/Wallet/Primitive/AddressDerivation.hs#L708-L711

and then the `Mainnet` implementation always uses `networkId == 1` as part of the tag:

```haskell
instance PaymentAddress 'Mainnet ShelleyKey 'CredFromKeyK where
    paymentAddress paymentK = do
        Address $ BL.toStrict $ runPut $ do
            putWord8 (enterprise + networkId)
            putByteString . blake2b224 . xpubPublicKey . getKey $ paymentK
      where
        enterprise = 96
        networkId = 1
```

https://github.com/input-output-hk/cardano-wallet/blob/master/lib/wallet/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs#L287

Reply:
aha, and you are right! Mainnet+Staging -> 1, the rest -> 0. I will correct that. Thanks

Reply:
done